### PR TITLE
Add osx to testing environments

### PR DESCRIPTION
**Pull request by kagharpure**
Using the osx image's default python3.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -8,8 +8,11 @@ python:
   - 3.7
 
 env:
-  - KARTOTHEK_ARROW_VERSION=0.12.1
```
Comment (fjetter) [inline .travis.yml]: Instead of relying on the ordering here you should define the environemnt for the OSX job explicitly. I believe you can provide an `env` key which will overwrite this global setting

File: .travis.yml [hunk 2]
```diff
@@ -8,8 +8,11 @@ python:
...
   - KARTOTHEK_ARROW_VERSION=0.13.0
+  - KARTOTHEK_ARROW_VERSION=0.12.1
+
+before_cache: rm -f "$HOME/.cache/pip/log/debug.log"
```
Comment (fjetter) [inline .travis.yml]: Is this caching actually required or a nice to have?

File: .travis.yml [hunk 3]
```diff
@@ -35,6 +38,29 @@ after_success:
...
+  include:
+    - name: "osx + builtin python3"
+      os: osx
+      language: sh
+      before_cache:
+              - rm -f "$HOME/Library/Caches/pip/log/debug.log"
+      cache:
+        directories:
+          - "$HOME/Library/Caches/pip"
+      before_install:
```
Comment (fjetter) [inline .travis.yml]: What's different to the other jobs? Why do we need to define this step again. The individual steps look pretty familiar

File: .travis.yml [hunk 4]
```diff
@@ -35,6 +38,29 @@ after_success:
   - pip install codecov
   - codecov
 
+matrix:
+  allow_failures:
+    - os: osx
+  include:
+    - name: "osx + builtin python3"
+      os: osx
+      language: sh
```
Comment (fjetter) [inline .travis.yml]: Language is python. why specify as shell?

File: .travis.yml (hunk 2 above)
Comment (fjetter) [reply .travis.yml]: The build times look worse, this is why I'm asking. This might, of course, be unrelated and simply due to high load on travis but if it's related to this caching, I'd rather not use it

Comment (kagharpure) [reply .travis.yml]: a) Good to have
b) I tried a bunch of runs last night - I think there was a difference of ~1 minute; therefore a) good to have. Can be removed.

File: .travis.yml (hunk 3 above)
Comment (kagharpure) [reply .travis.yml]: The cache location is different on osx and simply setting `cache: pip` didn't seem to work as expected for me (I did some searching, I think this is due to Travis not officially supporting python on osx).

Comment (fjetter) [reply .travis.yml]: I was referring to everything else than the cache

File: .travis.yml (hunk 1 above)
Comment (kagharpure) [reply .travis.yml]: Fair point, changing.

File: .travis.yml (hunk 4 above)
Comment (kagharpure) [reply .travis.yml]: :D 

tldr: doesn't work with osx. 

Long version: when you specify `language: python` along with `os: osx`, it tries to download the specified python version from a location that doesn't have it and the build errors out. I was trying some complicated mess of an approach earlier on with homebrew (where I was facing some issues with cache-ing) but I realized early yesterday that the osx images on azure pipelines already comes with a version of `python3` (3.6.5), so I am using that directly now.

File: .travis.yml (hunk 2 above)
Comment (fjetter) [reply .travis.yml]: From what I can see the build takes longer than before.

**Comment by kagharpure** [comment]
> You mentioned issues with homebrew earlier. I don't see you interacting with homebrew

tldr: homebrew issues aren't relevant anymore.

Long version: when you specify `language: python` along with `os: osx`, it tries to download the specified python version from a location that doesn't have it and the build errors out. I was trying some complicated mess of an approach earlier on with homebrew (where I was facing some issues with cache-ing) but I realized early yesterday that the osx images on azure pipelines already comes with a version of python3 (3.6.5), so I am using that directly now.

File: .travis.yml (hunk 3 above)
Comment (kagharpure) [reply .travis.yml]: Right. So: 

a) `before_install` doesn't stack
b) we need to explicitly setup a virtual environment in case of osx (which happens by default for linux), so the common steps need essentially to be redefined.

File: .travis.yml (hunk 2 above)
Comment (kagharpure) [reply .travis.yml]: Might be because of load; unsure. The time saving I observed at night was ~1 minute so not that big of a difference. Removing.

File: .travis.yml (hunk 3 above)
Comment (kagharpure) [reply .travis.yml]: And splitting it out into a common shell script doesn't work as expected. For example: https://travis-ci.com/kagharpure/kartothek/jobs/219631966

Comment (kagharpure) [reply .travis.yml]: Actually, hang on; I think sourcing the script ought to fix this because it's then not running it in a separate shell - just checking now.

Comment (fjetter) [reply .travis.yml]: The linked build fails because the requirements pinning wasn't successful. not because the build script was wrong.

I'm wondering if the setup of the virtualenv can't be used for all builds. I don't see a reason why mac is special here. The same script should run on linux as well

Comment (kagharpure) [reply .travis.yml]: > The linked build fails because the requirements pinning wasn't successful. not because the build script was wrong.

That's because it's using the system `pip`, not the pip updated in the venv created above.

 > 'm wondering if the setup of the virtualenv can't be used for all builds. I don't see a reason why mac is special here. The same script should run on linux as well

If the os is linux, travis creates a venv by default, the user needn't explicitly specify that it be created.

File: .travis.yml (hunk 4 above)
Comment (crepererum) [reply .travis.yml]: > Long version: when you specify language: python along with os: osx, it tries to download the specified python version from a location that doesn't have it and the build errors out.

That sounds like a travis bug to me. Is there an upstream issue for that? If not, can you please create one? In either case, we need a link to that issue as a comment in the yml file.

Comment (kagharpure) [reply .travis.yml]: It's called out fairly clearly here: https://docs.travis-ci.com/user/languages/python/

**Comment by kagharpure** [comment]
@fjetter, @crepererum - anything left to do before we can consider this merge-ready?

**Comment by kagharpure** [comment]
@fjetter / @crepererum - can either of you merge please? I am not authorized to do so.

**Comment by lr4d** [comment]
@kagharpure Do you know why we are getting the segfault during the OS X tests?

**Comment by fjetter** [comment]
@lr4d see https://github.com/JDASoftwareGroup/kartothek/issues/54

The fix should be trivial

**Comment by lr4d** [comment]
Thanks, @fjetter . I see it has been fixed upstream so we might as well wait for the next pyarrow release
